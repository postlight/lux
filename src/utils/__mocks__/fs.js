/* @flow */

import MemoryFileSystem from 'memory-fs'

const fs = new MemoryFileSystem()

export const reset = () => {
  fs.data = {}
}

export const stat = jest.fn().mockImplementation((...args) => {
  fs.stat(...args)
})

export const readdir = jest.fn().mockImplementation((...args) => {
  fs.readdir(...args)
})

export const rmdir = jest.fn().mockImplementation((...args) => {
  fs.rmdir(...args)
})

export const unlink = jest.fn().mockImplementation((...args) => {
  fs.unlink(...args)
})

export const readlink = jest.fn().mockImplementation((...args) => {
  fs.readlink(...args)
})

export const mkdir = jest.fn().mockImplementation((...args) => {
  fs.mkdir(...args)
})

export const readFile = jest.fn().mockImplementation((...args) => {
  fs.readFile(...args)
})

export const exists = jest.fn().mockImplementation((...args) => {
  fs.exists(...args)
})

export const writeFile = jest.fn().mockImplementation((...args) => {
  fs.writeFile(...args)
})
