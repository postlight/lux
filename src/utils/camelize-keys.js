import { camelize } from 'inflection';

const { entries } = Object;
const { isArray } = Array;

export default function camelizeKeys(obj, deep = false) {
  return isArray(obj) ? obj.slice() : entries(obj)
    .reduce((result, [key, value]) => {
      if (deep && value && typeof value === 'object' && !isArray(obj)) {
        value = camelizeKeys(value, true);
      }

      return {
        ...result,
        [camelize(key.replace(/-/g, '_'), true)]: value
      };
    }, {});
}
