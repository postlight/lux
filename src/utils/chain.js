// @flow
import type { Chain } from '../interfaces';

/**
 * @private
 */
export default function chain<T>(source: T): Chain<T> {
  return {
    pipe<U>(handler: (value: T) => U): Chain<U> {
      return chain(handler(source));
    },

    value(): T {
      return source;
    }
  };
}
