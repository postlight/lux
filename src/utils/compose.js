// @flow

/**
 * @private
 */
export function tap<T>(input: T): T {
  console.log(input); // eslint-disable-line no-console
  return input;
}

/**
 * @private
 */
export function compose<T>(...funcs: Array<(input: T) => T>): (input: T) => T {
  return input => funcs.reduceRight((value, fn) => fn(value), input);
}

/**
 * @private
 */
export function composeAsync<T>(
  ...funcs: Array<(input: T) => T | Promise<T>>
): (input: T) => Promise<T> {
  return input => funcs.reduceRight(
    (value, fn) => Promise.resolve(value).then(fn),
    Promise.resolve(input)
  );
}
