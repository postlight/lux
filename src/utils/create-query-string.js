// @flow
import entries from './entries';

/**
 * A replacement for querystring.stringify that supports nested objects.
 *
 * @private
 */
export default function createQueryString(
  source: Object,
  property?: string
): string {
  return entries(source).reduce((result, [key, value], index, arr) => {
    if (index > 0) {
      result += '&';
    }

    if (property) {
      result += (
        property
        + encodeURIComponent('[')
        + key
        + encodeURIComponent(']')
        + '='
      );
    } else {
      result += `${key}=`;
    }

    if (value && typeof value === 'object') {
      if (Array.isArray(value)) {
        result += value.map(encodeURIComponent).join();
      } else {
        result = (
          result.substr(0, result.length - (key.length + 1))
          + createQueryString(value, key)
        );
      }
    } else if (!value && typeof value !== 'number') {
      result += 'null';
    } else {
      result += encodeURIComponent(value);
    }

    return result;
  }, '');
}
