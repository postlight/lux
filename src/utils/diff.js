// @flow

/**
 * @private
 */
export function map<K, V>(a: Map<K, V>, b: Map<K, V>): Map<K, V> {
  return Array
    .from(b)
    .reduce((result, [key, value]) => {
      if (a.get(key) !== value) {
        result.set(key, value);
      }

      return result;
    }, new Map());
}
