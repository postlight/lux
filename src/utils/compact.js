/* @flow */

import { isNil } from '@lux/utils/is-type'

const compact = <T>(source: Array<T>): Array<T> =>
  source.filter(value => !isNil(value))

export default compact
