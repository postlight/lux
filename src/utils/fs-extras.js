/* @flow */

import * as path from 'path'

import * as fs from 'mz/fs'

import * as flatten from '@lux/utils/flatten'

/**
 * Recursively remove files and directories if they exist. YOLO.
 */
const rmrf = (target: string): Promise<boolean> =>
  fs
    .stat(target)
    .then(
      stats =>
        stats.isDirectory()
          ? fs
              .readdir(target)
              .then(files => files.map(file => path.join(target, file)))
              .then(files => files.map(rmrf))
              .then(promises => Promise.all(promises))
              .then(() => fs.rmdir(target))
          : fs.unlink(target),
    )
    .then(() => true)
    .catch(err => {
      if (err.code === 'ENOENT') {
        return true
      }
      return err
    })

/**
 * Determine if a file in a directory matches a pattern.
 */
const existsInDir = (dir: string, pattern: RegExp): Promise<boolean> =>
  fs.readdir(dir).then(files => files.some(file => pattern.test(file)))

/**
 * Recursively read a directory.
 */
const readdirRec = (
  target: string,
  stripPath?: boolean = true,
): Promise<Array<string>> =>
  fs
    .readdir(target)
    .then(files =>
      files
        .map(file => path.join(target, file))
        .map(file => fs.stat(file).then(stats => [file, stats])),
    )
    .then(promises => Promise.all(promises))
    .then(files =>
      files.map(([file, stats]) => {
        if (stats.isDirectory()) {
          return readdirRec(file, false)
        }
        return file
      }),
    )
    .then(promises => Promise.all(promises))
    .then(flatten.shallow)
    .then(
      files =>
        stripPath ? files.map(file => path.relative(target, file)) : files,
    )

/**
 * Recursively create a directory path.
 */
const mkdirRec = (target: string): Promise<void> =>
  fs
    .mkdir(path.dirname(target))
    .catch(err => {
      if (err.code === 'ENOENT') {
        return mkdirRec(path.dirname(target))
      }
      return err
    })
    .then(() => fs.mkdir(target))

/**
 * Higher order function that returns a function that determins if a
 * file has a matching extension.
 */
const isExt = (ext: string): ((file: string) => boolean) => file =>
  path.extname(file) === ext

export { existsInDir, isExt, mkdirRec, readdirRec, rmrf }
