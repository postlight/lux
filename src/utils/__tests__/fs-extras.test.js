/* @flow */

import * as mockFs from 'fs'
import * as path from 'path'

import * as fs from 'mz/fs'

import * as fse from '@lux/utils/fs-extras'

jest.mock('fs')

// $FlowFixMe
afterEach(mockFs.reset)

describe('rmrf()', () => {
  const root = path.join(path.sep, 'x', 'y', 'z')
  const file = path.join(root, 'test.json')

  beforeEach(async () => {
    await fse.mkdirRec(root)
    await fs.writeFile(file, '{}')
  })

  afterEach(async () => {
    await fse.rmrf(root)
  })

  test('it can remove a file', async () => {
    expect(await fse.rmrf(file)).toBe(true)
    expect(await fs.exists(file)).toBe(false)
  })

  test('it can remove a directory and its contents', async () => {
    expect(await fse.rmrf(root)).toBe(true)
    expect(await fs.exists(root)).toBe(false)
  })

  test('it does not fail if the target does not exist', async () => {
    const target = path.join(root, 'not-a-file')

    expect(await fse.rmrf(target)).toBe(true)
  })

  test('valid errors bubble up', async () => {
    await fse.rmrf('test').catch(err => {
      expect(err).toEqual(expect.any(Error))
    })
  })
})

test('existsInDir()', async () => {
  const root = path.join(path.sep, 'test')

  await fs.mkdir(root)
  await fs.writeFile(path.join(root, 'package.json'), '{}')

  expect(await fse.existsInDir(root, /\.json$/)).toBe(true)
})

test('readdirRec()', async () => {
  await fse.mkdirRec(path.join(path.sep, 'x', 'y', 'z'))
  await Promise.all([
    fs.writeFile(path.join(path.sep, 'x', 'x.txt'), 'test x'),
    fs.writeFile(path.join(path.sep, 'x', 'y', 'y.txt'), 'test y'),
    fs.writeFile(path.join(path.sep, 'x', 'y', 'z', 'z.txt'), 'test z'),
  ])

  expect(await fse.readdirRec(path.join(path.sep, 'x'))).toMatchSnapshot()
})

describe('mkdirRec()', () => {
  test('it can recursively create a directory', async () => {
    const target = path.join(path.sep, 'x', 'y', 'z')

    expect(await fs.exists(target)).toBe(false)

    await fse.mkdirRec(target)
    expect(await fs.exists(target)).toBe(true)
  })

  test('valid errors still bubble up', async () => {
    await fse.mkdirRec('test').catch(err => {
      expect(err).toEqual(expect.any(Error))
    })
  })
})

test('isExt()', () => {
  const isJSFile = fse.isExt('.js')

  expect(isJSFile('test.js')).toBe(true)
  expect(isJSFile('test.json')).toBe(false)
})
