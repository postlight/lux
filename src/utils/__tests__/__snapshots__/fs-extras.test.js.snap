// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`readdirRec() 1`] = `
Array [
  "y/z/z.txt",
  "y/y.txt",
  "x.txt",
]
`;
